[Codegen] Expose namespace argument for static dispatch

larryliu0820 opened the pull request:
For static dispatch we are hardcoding namespace to be `at` for backend-specific C++ functions, e.g., `at::cpu::add()`. We are extending it to accept namespaces from callsite. This is a temporary solution, in the long run we want to introduce custom namespace into codegen system, e.g., we should be able to add `at::` to `native_functions.yaml` and parse it into `NativeFunction`. This needs a bit more design.

ezyang reviewed (approved):
Please do the long term thing lol

larryliu0820 commented:
> Please do the long term thing lol

Will do in next PR

larryliu0820 committed:
[executorch] ATen integration: support different namespaces for ATen & Executorch C++ APIs (#77710)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/77710

This is an important step to enable ATen integration into Executorch. This diff lets generated `Functions.h` either call `at::<dispatch_key>::<ops>` or `torch::executor::function::<dispatch_key>::<ops>` based on the BUCK config. `at::<dispatch_key>::<ops>` symbols can be found in `caffe2/gen_aten` target, so if we want to use ATen kernels we need to depend on that target. This way we can switch between ATen kernels and portable kernels.

See the following diagram:

{F733113492}

Generated `Functions.h` code snippet:
use_aten_lib=False:
```
#include "CPUFunctions.h"
#include "CompositeImplicitAutogradFunctions.h"
#include "CompositeExplicitAutogradFunctions.h"

namespace torch {
namespace executor {
namespace function {

// aten::add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
TORCH_API inline at::Tensor add(const at::Tensor & self, const at::Tensor & other, const at::Scalar & alpha=1) {
    return torch::executor::function::cpu::add(self, other, alpha);
}

} // namepsace function
} // namespace executor
} // namespace torch

```

use_aten_lib=True:
```
#include <ATen/CPUFunctions.h>
#include <ATen/CompositeImplicitAutogradFunctions.h>
#include <ATen/CompositeExplicitAutogradFunctions.h>

namespace torch {
namespace executor {
namespace function {

// aten::add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
TORCH_API inline at::Tensor add(const at::Tensor & self, const at::Tensor & other, const at::Scalar & alpha=1) {
    return at::cpu::add(self, other, alpha);
}

} // namepsace function
} // namespace executor
} // namespace torch
```

Right now it's still a hard switch between ATen and portable kernels, but in the future we should be able to generate per-operator namespace.

Test Plan:
```
[ (5a33fe6ed|remote/master) | devvm1625 ~/fbsource] buck test //xplat/executorch/executor:executor_test
Building: finished in 1.4 sec (100%) 43/43 jobs, 0/43 updated
  Total time: 1.4 sec
Testing: finished in 0.5 sec (15 PASS/0 FAIL)
BUILD SUCCEEDED
RESULTS FOR //xplat/executorch/executor:executor_test
PASS    <100ms 15 Passed   0 Skipped   0 Failed   //xplat/executorch/executor:executor_test
TESTS PASSED
```

Reviewed By: ezyang, iseeyuan

Differential Revision: D36466976

fbshipit-source-id: 70c2477c01628781ff0fb4c50f40f9ad76243fc7
